modules/common/tls/tls.go

```diff
@@ -0,0 +1,235 @@
+		service.SecretName = "default-tls-secret"
+	} else {
+		err := EnsureOrCreateSecret(clientset, namespace, service.SecretName, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if ca.CaSecretName == "" {
+		ca.CaSecretName = "default-ca-secret"
+	} else {
+		err := EnsureOrCreateSecret(clientset, namespace, ca.CaSecretName, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &TLS{
+		Service: service,
+		Ca:      ca,
+	}, nil
+}
+
+// EnsureOrCreateSecret - checks if a secret with the given name exists in the specified namespace
+func EnsureOrCreateSecret(clientset *kubernetes.Clientset, namespace, secretName string, data map[string][]byte) error {
+	_, err := clientset.CoreV1().Secrets(namespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Secret not found, create it
+			secret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretName,
+					Namespace: namespace,
+				},
+				Data: data, // The data to store in the secret - the actual data provided by cert-manager formatted as "tls.crt": []byte("..."),  "tls.key": []byte("..."),
+			}
+			_, err := clientset.CoreV1().Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+			if err != nil {
+				return fmt.Errorf("Error creating secret %s: %w", secretName, err)
+			}
+		} else {
+			return fmt.Errorf("Error validating secret %s: %w", secretName, err)
+		}
+	}
+	return nil
+}
+
+// CreateVolumeMounts - add volume mount for TLS certificate and CA certificates
+func CreateVolumeMounts(tls *TLS) []corev1.VolumeMount {
+	var volumeMounts []corev1.VolumeMount
+
+	if tls.Service.SecretName != "" {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      "tls-certs",
+			MountPath: "/etc/tls-certs",
+			ReadOnly:  true,
+			SubPath:   "tls.crt",
+		},
+			corev1.VolumeMount{
+				Name:      "tls-certs",
+				MountPath: "/etc/tls-certs",
+				ReadOnly:  true,
+				SubPath:   "tls.key",
+			})
+	}
+
+	if tls.Ca.CaSecretName != "" {
+		for i := 0; i < MaxNumberOfCAs; i++ {
+			volumeMounts = append(volumeMounts, corev1.VolumeMount{
+				Name:      fmt.Sprintf("ca-cert-%d", i),
+				MountPath: fmt.Sprintf("/etc/ca-certs/%d", i),
+				ReadOnly:  true,
+				SubPath:   fmt.Sprintf("ca-%d.crt", i),
+			})
+		}
+	}
+
+	return volumeMounts
+}
+
+// CreateVolumes - add volume for TLS certificate and CA certificates
+func CreateVolumes(tls *TLS) []corev1.Volume {
+	var volumes []corev1.Volume
+
+	if tls.Service.SecretName != "" {
+		volumes = append(volumes, corev1.Volume{
+			Name: "tls-certs",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: tls.Service.SecretName,
+					Items: []corev1.KeyToPath{
+						{
+							Key:  "tls.crt",
+							Path: "tls.crt",
+						},
+						{
+							Key:  "tls.key",
+							Path: "tls.key",
+						},
+					},
+				},
+			},
+		})
+	}
+
+	if tls.Ca.CaSecretName != "" {
```

Comment:
My understanding is that we plan to support two high level use cases:
 1. we create certificates for the control plane services, so we need one CA cert to verify all those certificates  when doing TLS communication between pods.
 2. we need to support additional CA in case specific services require to connect to endpoints for which we don't create certificate.

Given that we always need 1. for TLS, and 2. seems to be unbounded, could we expose that in the signature of CreateVolumes, like maybe e.g.
```
CreateVolumes(tls *TLSService, ca []TLSCa)
``` 

This way we wouldn't need MaxNumberOfCAs

Reply:
Yes, in general we need multiple CA cert - one for validate the certs used between the pods, meaning the tls.crt certs. Other CA certs can be used to validate certificates provided by external clients.

So yes, you concluded it correctly in points 1. and 2. And there's no issue with using:
`CreateVolumes(tls *TLSService, ca []TLSCa)`